hbase_storage/src/main/java/com/splicemachine/storage/HScan.java

```diff
@@ -43,6 +48,57 @@ public boolean isDescendingScan(){
         return scan.isReversed();
     }
 
+    /**
+     * Take a list of [startKey, stopKey) rowkey pairs, where the stopKey is excluded,
+     * convert it to the corresponding {@link MultiRowRangeFilter} and attach
+     * it to the {@link Scan} held in this {@link HScan} as a {@link Filter}.
+     *
+     * @param rowkeyPairs the list of [startKey, stopKey) pairs to convert.
+     *
+     * @return if this {@link HScan} currently has no filter, build a new
+     * {@link MultiRowRangeFilter} out of {@code rowkeyPairs} and attach it
+     * as a filter.  If this {@link HScan} already has a {@link MultiRowRangeFilter},
+     * replace it with a new MultiRowRangeFilter built from {@code rowkeyPairs}.
+     * If {@code rowkeyPairs} has no elements, do not build a filter.
+     *
+     * @throws IOException
+     *
+     * @notes A possible future enhancement is, instead of replacing an old
+     * MultiRowRangeFilter with a new one, build the intersection of the
+     * old filter and the new one.
+     *
+     * @see     Scan
+     */
+    @Override
+    public void addRowkeyRangesFilter(List<Pair<byte[],byte[]>> rowkeyPairs) throws IOException {
+        if (rowkeyPairs == null || rowkeyPairs.size() < 1) {
+            return;
+        }
+        Filter currentFilter = scan.getFilter();
+        if (currentFilter != null && !(currentFilter instanceof MultiRowRangeFilter))
+            return;
+
+        List<MultiRowRangeFilter.RowRange> ranges = new ArrayList<>(rowkeyPairs.size());
+        byte[] startKey;
+        byte[] stopKey;
+        for (Pair<byte[],byte[]> startStopKey: rowkeyPairs) {
+            startKey=startStopKey.getFirst();
+            stopKey=startStopKey.getSecond();
+            MultiRowRangeFilter.RowRange rr =
+            new MultiRowRangeFilter.RowRange(startKey, true,
+                                             stopKey, false);
+            ranges.add(rr);
+        }
+
+        if (ranges.size() > 0) {
+            MultiRowRangeFilter filter = new MultiRowRangeFilter(ranges);
+            scan.setFilter(filter);
+            List<MultiRowRangeFilter.RowRange> sortedRanges = filter.getRowRanges();
+            scan.withStartRow(sortedRanges.get(0).getStartRow());
```

Comment:
The original start key of 5 is never getting restored, instead the scan reads row (1,6,6), then calls Scans.qualifyRecordFromRow to apply the a2>=5 predicate which was built into a qualifier, filtering out the row.  It would be more efficient to use the more restrictive start key.  In HScan.addRowkeyRangesFilter, I will check the pre-existing start and stop keys and only replace them if the key from the filter is more restrictive.

Reply:
Yes, it would be good to honor the more restrictive start/stop key. Thanks!